---
title: "Basic Bioinformatics Report"
author:
    - "CHOP Microbiome Center"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
    html_document:
        highlight: tango
        number_sections: no
        theme: default
        toc: yes
        toc_depth: 4
        toc_float:
            collapsed: no
            smooth_scroll: yes
---

```{r, echo = FALSE}
### ================
###   knitr setup
### ================

library(knitr)
opts_chunk$set(
  tidy = FALSE,
  cache = FALSE,
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  dpi = 100,
  fig.align = "center",
  fig.width = 6,
  fig.height = 6
  )
```

```{r, message = FALSE}
### ================
###   R packages
### ================

library(tidyverse)
library(qiimer)
library(pander)
library(vegan)
library(pheatmap)
```

```{r}
heat <- function(s_Heat, props_Heat, grps, satu_limit = satu_limit, prop_cut = prop_cut, 
                 annotation_legend = T,
                 must_add_row = NULL,
                 gaps_col = NULL,
                 gaps_row = NULL,
                 show_colnames = T,
                 use_annotation = T,
                 annotation_colors = NA) {
  
  s_Heat <- s_Heat %>%
    dplyr::select(SampleID, grps) %>%
    unique() %>%
    arrange_(.dots = grps)
  
  anno <- s_Heat %>% select(-SampleID)
  rownames(anno) <- s_Heat$SampleID
  colnames(anno) <- grps
  
  props_Heat <- props_Heat[, s_Heat$SampleID]
  row_to_show <- rownames(props_Heat)[apply(props_Heat, 1, max) >= prop_cut]
  row_to_show <- c(row_to_show, must_add_row)
  
  props_Heat <- props_Heat[row_to_show, rownames(anno)]
  
  color = saturated_rainbow(101, saturation_limit = satu_limit)
  breaks = c(0, 1e-10, seq(0.001, 1, length.out = 100))
  
  if (use_annotation) {
    g <- pheatmap(props_Heat, annotation = anno, cluster_cols = F, cluster_rows = F, 
                  color = color, breaks = breaks, annotation_colors = annotation_colors,
                  gaps_col = gaps_col, gaps_row = gaps_row,
                  show_colnames = show_colnames, annotation_legend = annotation_legend, 
                  cellwidth = 8, cellheight = 8, fontsize_col = 8, fontsize_row = 8)
  } else {
    g <- pheatmap(props_Heat, cluster_cols = F, cluster_rows = F, 
                  color = color, breaks = breaks, annotation_colors = annotation_colors,
                  gaps_col = gaps_col, gaps_row = gaps_row,
                  show_colnames = show_colnames, annotation_legend = annotation_legend, 
                  cellwidth = 8, cellheight = 8, fontsize_col = 8, fontsize_row = 8)
  }
  
  return(g)
}
```

```{r}
### =====================
###   define constants
### =====================

### minimum QC read count threshold
min_reads <- 5

### rarefying subsample size 
richness_subsample_size <- 20

### mapping file path
mapping_file_fp <- snakemake@input[["mapping_fp"]]

### demux counts file path
demux_count_fp <- snakemake@input[["demux_count_fp"]]

### otu table file path
feature_table_fp <- snakemake@input[["feature_table_fp"]]

### taxonomic assignment file path
taxo_assignment_fp <- snakemake@input[["taxo_assignment_fp"]]

### unweighted UniFrac file path
#uu_fp <- ""

### weighted UniFrac file path
#wu_fp <- ""

### Faith phylogenetic diversity
faith_fp <- snakemake@input[["faith_fp"]]

### study group variable
study_group_var <- snakemake@config[["report"]][["study_group_var"]]
```

```{r, warning = F}
### ==================
###   read OTU table and taxonomy assignment
### ==================

### read otu count data
counts <- read_tsv(file = feature_table_fp, skip = 1) %>%
  column_to_rownames(var = "#OTU ID") %>%
  as.matrix()

### taxonomy assignment
ta <- read_tsv(file = taxo_assignment_fp) %>%
  mutate(Taxon = str_remove(Taxon, "(; [kpcofgs]__)+$")) %>%
  column_to_rownames(var = "Feature ID")
ta <- ta[rownames(counts), ] # make sure that OTU table and taxonomy assignment have the same rownames

### taxonomy assignment as a data frame
md <- ta %>%
  pull(Taxon)
names(md) <- rownames(ta)
adf <- split_assignments(md) 
```

```{r}
### ==================
###   count data
### ==================

### get read counts after demultiplexing
demux <- read_csv(file = demux_count_fp) %>%
  setNames(c("SampleID", "demux_Read_Counts"))

### get read counts after denosing by DADA2 in QIIME2 pipeline
denoise <- colSums(counts) %>% 
  as.data.frame() %>%
  rownames_to_column() %>%
  setNames(c("SampleID", "denoise_Read_Counts"))

### get read counts after removing contamination
is_mitochondrial <- grepl("mitochondria", adf$Family)
is_chloroplast <- grepl("Chloroplast", adf$Class)
is_unassigned <- grepl("Unassigned", adf$Kingdom)
is_archaea <- grepl("Archaea", adf$Kingdom)
is_contam <- is_mitochondrial | is_chloroplast | is_unassigned 

counts <- counts[!is_contam, ]
adf <- adf[!is_contam, ]

qc <- colSums(counts) %>% 
  as.data.frame() %>%
  rownames_to_column() %>%
  setNames(c("SampleID", "QC_Read_Counts"))
```

```{r}
### ==================
###   alpha diversity
### ==================

richness <- rarefy(t(counts), richness_subsample_size) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  setNames(c("SampleID", "Richness"))

shannon <- diversity(t(counts)) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  setNames(c("SampleID", "Shannon"))

faith <- read_tsv(file = faith_fp) %>%
  setNames(c("SampleID", "Faith"))
```  

```{r}
### ==================
###   read mapping file and add count / alpha diversity info
### ==================

### read mapping file
s0 <- read_qiime_mapping_file(mapping_file_fp) 

### add read counts data
s <- s0 %>%
  left_join(demux, by = "SampleID") %>%
  left_join(denoise, by = "SampleID") %>%
  left_join(qc, by = "SampleID") %>%
  mutate(above_min_reads = QC_Read_Counts > min_reads) %>%
  mutate(QC_read_call = factor(ifelse(above_min_reads, "above threshold", "below threshold")))

### add alpha diversity measures
s <- s %>%
  left_join(richness, by = "SampleID") %>%
  left_join(shannon, by = "SampleID") %>%
  left_join(faith, by = "SampleID")
```

```{r}
### ===========================
###   general data preprocessing and cleaning
### ===========================

### determine lab control samples
s$is_control <- grepl("emptywell|extractblank|dnafree|geneblock", s$SampleID, ignore.case = T)
experimental_sample_list <- s$SampleID[!grepl("emptywell|extractblank|dnafree|geneblock", s$SampleID, ignore.case = T)]

### possible issue 1: any mismatch between s$SampleID and colames(counts)?
in_s_not_in_counts <- setdiff(s$SampleID, colnames(counts))
in_s_not_in_counts <- unique(c(in_s_not_in_counts, s$SampleID[!is.na(s$QC_Read_Counts) & s$QC_Read_Counts == 0]))
#if (length(in_s_not_in_counts) > 0) stop (simpleError("These SampleID(s) are in the mapping file, but not found in the feature table.", paste(in_s_not_in_counts, collapse=" ")))

in_counts_not_in_s <- setdiff(colnames(counts), s$SampleID)
#if (length(in_counts_not_in_s) > 0) stop (simpleError("These SampleID(s) are in the feature table, but not found in the mapping file.", paste(in_counts_not_in_s, collapse=" ")))

common_sample_list <- s$SampleID[!(s$SampleID %in% in_s_not_in_counts)]
s <- s[s$SampleID %in% common_sample_list, ]
```

# Sample size

```{r}
s_Table <- addmargins(table(s[, study_group_var], s$QC_read_call))
pander(s_Table, caption = "Samples with enough read counts by study group")
```

# Taxonomic heatmap

```{r}
prop_cut <- 0.01
satu_limit <- 0.4
```

Each column of the heatmap represents one sample and each row represents one taxon, typically a genus. Taxa were included in the chart if the abundance in any sample exceeded `r 100*prop_cut`%. 

The chart is colored white if taxa were not observed in the sample, dark blue if taxa were observed at very low abundance. This allows the reader to quickly survey presence/absence. Abundance values exceeding `r 100*satu_limit`% are colored red, indicating an extremely dominant species.

```{r, fig.width = 10, fig.height = 7, out.width = "90%"}
s_Heat <- s 
cts_Heat <- counts[, s_Heat$SampleID]  
a <- simplify_assignments(adf, rank1 = "Phylum", rank2 = "Genus")
summed_cts <- rowsum(cts_Heat, a) 
summed_props <- sweep(summed_cts, 2, colSums(summed_cts), "/") 
grps <- study_group_var
heat(s_Heat, summed_props, grps = grps, satu_limit = satu_limit, prop_cut = prop_cut)
```

# Alpha diversity

Alpha diversity was assessed by the expected number of observed OTUs (out of rarefying sample size of `r richness_subsample_size`), Shannon index, and Faithâ€™s phylogenetic diversity.

```{r}
s_Alpha <- s %>%
  select(SampleID, Richness, Shannon, Faith) %>%
  gather(key = alpha_measure, value = alpha_value, -SampleID) %>%
  mutate(alpha_measure = fct_relevel(alpha_measure, "Richness", "Shannon", "Faith")) %>%
  filter(!is.na(alpha_value)) %>%
  left_join(s, by = "SampleID") %>%
  droplevels()
```

```{r, fig.width = 8, fig.height = 3}
g <- s_Alpha %>%
  ggplot(aes(x = get(study_group_var), y = alpha_value)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter(width = 0.2) +
  ylim(0, NA) +
  theme(aspect.ratio = 1) +
  #theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  labs(x = NULL) +
  facet_wrap(~alpha_measure, ncol = 3, scales = "free_y")
print(g)
```

# Appendix: read counts per sample

__demux_Read_Counts__: read counts after demultiplexing 

__denoise_Read_Counts__: read counts after the denoising step in QIIME2 pipeline

__QC_Read_Counts__: read counts after removing contamination (mitochondria, chloroplast, or Unassigned)

```{r}
s %>% select(SampleID, demux_Read_Counts, denoise_Read_Counts, QC_Read_Counts) %>%
  arrange(desc(QC_Read_Counts)) %>%
  pander()
```
